### PR TITLE
JS: improve support for destructuring and default values

### PR DESCRIPTION
**Pull request by esben-semmle**
This PR contains three related, but independent changes, all motivated by <https://jira.semmle.com/browse/ODASA-7715>.

The first change is a def-use bugfix that removes the direct step from `x` to `y` in `var { y } = x;`, the implicit step from `x.y` to `y` is still performed in the dataflow graph.

The other changes add support for default values in parameters and destructuring assignments. The type inference already had support for default parameter values.

Please note that the introduction of default parameter values has the effect that parameters now have a predecessor in the dataflow graph, which may be surprising, as seen in the `RangeAnalysis.qll` change.

Implementationwise, there is a small nit: unlike default parameter values, default values in destructuring assignments can not be implemented smoothly in the def-use graph at the moment, so they are implemented in the dataflow graph where destructuring are handled already. The reason for this is that a single `VarDef` represents multiple sources and targets, and the relation between each source and target is lost when `VarDef::getSource()` is used.

A larger evaluation is on the way, but a [preliminary evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/defuse-default_1549486180808) shows plenty of result improvements at no performance cost.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/DefUse.qll [hunk 1]
```diff
@@ -182,14 +182,26 @@ class VarDef extends ControlFlowNode {
   /** Gets a variable defined by this node, if any. */
   Variable getAVariable() { result = getTarget().(BindingPattern).getAVariable() }
 
+  private AST::ValueNode getTargetSource(Expr target) { defn(this, target, result) }
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/DefUse.qll]: I find the name of this predicate very confusing. Could we just inline it into its two uses?

File: javascript/ql/src/semmle/javascript/DefUse.qll [hunk 2]
```diff
@@ -182,14 +182,26 @@ class VarDef extends ControlFlowNode {
...
+
   /**
    * Gets the source of this definition, that is, the data flow node representing
    * the value that this definition assigns to its target.
    *
    * This predicate is not defined for `VarDef`s where the source is implicit,
    * such as `for-in` loops or parameters.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/DefUse.qll]: Could you add a paragraph explaining that it's also not defined for destructuring definitions?

File: javascript/ql/src/semmle/javascript/DefUse.qll [hunk 3]
```diff
@@ -182,14 +182,26 @@ class VarDef extends ControlFlowNode {
...
-  AST::ValueNode getSource() { defn(this, _, result) }
+  AST::ValueNode getSource() {
+    exists(Expr target |
+      not target instanceof DestructuringPattern and getTargetSource(target) = result
+    )
+  }
+
+  /**
+   * Gets the source that this definition destructs, that is, the data flow node representing
+   * the value that contains a value that this definition assigns to its target.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/DefUse.qll]: This is very hard to understand. Could you please rephrase?

Also, I don't see anything in this predicate that restricts it to the destructuring case. It seems to be a strict superset of `getSource`.

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll [hunk 4]
```diff
@@ -987,6 +987,14 @@ module DataFlow {
     )
   }
 
+  /**
+   * Gets the data flow node representing the destructuring source of definition `def`, taking
+   * flow through IIFE calls into account.
+   */
+  private AST::ValueNode destructuringDefSourceNode(VarDef def) {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]: I'm not sure we need a separate predicate for this. The two uses of `defSourceNode/1` (in `localFlowStep` above and in `defSourceNode/2` below) look like they should work with both the destructuring and the non-destructuring case.

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll [hunk 5]
```diff
@@ -1010,15 +1010,19 @@ module DataFlow {
...
       or
       // handle destructuring assignments
-      exists(PropertyPattern pp | r = pp.getValuePattern() | result = TPropNode(pp))
+      exists(PropertyPattern pp | r = pp.getValuePattern() |
+        result = TPropNode(pp) or result = pp.getDefault().flow()
+      )
       or
       result = TElementPatternNode(_, r)
+      or
+      exists(ArrayPattern ap, int i | ap.getElement(i) = r and result = ap.getDefault(i).flow())
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll]: Do we have a test for this?

File: javascript/ql/src/semmle/javascript/RangeAnalysis.qll [hunk 6]
```diff
@@ -104,7 +104,9 @@ module RangeAnalysis {
   pragma[noinline]
   private predicate hasUniquePredecessor(DataFlow::Node node) {
     isRelevant(node) and
-    strictcount(node.getAPredecessor()) = 1
+    strictcount(node.getAPredecessor()) = 1 and
+    // exclude parameters with default values
+    not DataFlow::parameterNode(_) = node
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/RangeAnalysis.qll]: This now also excludes flow across local calls. Is that intentional?

File: javascript/ql/test/query-tests/Statements/UselessConditional/UselessConditionalGood.js [hunk 7]
```diff
@@ -99,3 +99,35 @@ function getLastLine(input) {
...
+			if (p) { // OK, whitelisted
+
+			}
+		});
+	}
+	h();
+});
+
+
+(function() {
```
Comment (xiemaisi) [inline javascript/ql/test/query-tests/Statements/UselessConditional/UselessConditionalGood.js]: What's this guy doing?

**Comment by esben-semmle** [comment]
All comments addressed, as a rebase, see https://github.com/Semmle/ql/compare/9b3381e0548d6887ebcb9333d22cccb638209012..5ad83360be73c21999425dfc3e8ac053817ad574 for the pretty diff. 

A big-apps.slugs [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/defuse-default_1549597590477) shows that this still scales. Do we want a full evaluation on default.slugs?

**Comment by xiemaisi** [comment]
LGTM. I would prefer a full evaluation, just because I've been caught out by unexpected corner cases in this area of the language before.

**Comment by esben-semmle** [comment]
(The full [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/defuse-default_1549726665920) is just more of what we already saw in the partial evaluation.)

**Pull request merged**